File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityProjection.scala
// tag::projection[]
package shopping.cart

import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.ShardedDaemonProcessSettings
import akka.cluster.sharding.typed.scaladsl.ShardedDaemonProcess
import akka.persistence.query.Offset
import akka.persistence.r2dbc.query.scaladsl.R2dbcReadJournal
import akka.projection.eventsourced.scaladsl.EventSourcedProvider
import akka.projection.scaladsl.SourceProvider
import akka.projection.{ProjectionBehavior, ProjectionId}
import shopping.cart.repository.ItemPopularityRepository
import akka.persistence.query.typed.EventEnvelope
import akka.projection.Projection
import akka.projection.r2dbc.scaladsl.R2dbcProjection

object ItemPopularityProjection {
  // tag::howto-read-side-without-role[]
  def init(
            system: ActorSystem[_],
            repository: ItemPopularityRepository): Unit = {

    implicit val sys = system

    def sourceProvider(sliceRange: Range): SourceProvider[Offset, EventEnvelope[ShoppingCart.Event]] =
      EventSourcedProvider
        .eventsBySlices[ShoppingCart.Event](
          system,
          readJournalPluginId = R2dbcReadJournal.Identifier,
          "ShoppingCart",
          sliceRange.min,
          sliceRange.max)

    def projection(sliceRange: Range): Projection[EventEnvelope[ShoppingCart.Event]] = {
      val minSlice = sliceRange.min
      val maxSlice = sliceRange.max
      val projectionId = ProjectionId("ShoppingCarts", s"carts-$minSlice-$maxSlice")

      R2dbcProjection
        .exactlyOnce(
          projectionId,
          settings = None,
          sourceProvider(sliceRange),
          handler = () =>
            new ItemPopularityProjectionHandler(s"carts-$minSlice-$maxSlice",
              system,
              repository))
    }

    ShardedDaemonProcess(system).initWithContext(
      name = "ShoppingCartProjection",
      initialNumberOfInstances = 4,
      behaviorFactory = { daemonContext =>
        val sliceRanges =
          EventSourcedProvider.sliceRanges(system, R2dbcReadJournal.Identifier, daemonContext.totalProcesses)
        val sliceRange = sliceRanges(daemonContext.processNumber)
        ProjectionBehavior(projection(sliceRange))
      },
      ShardedDaemonProcessSettings(system),
      stopMessage = ProjectionBehavior.Stop)
  }
}
// end::projection[]

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-java/src/main/java/shopping/cart/ItemPopularityProjection.java
// tag::projection[]
package shopping.cart;

import akka.actor.typed.ActorSystem;
import akka.cluster.sharding.typed.ShardedDaemonProcessSettings;
import akka.cluster.sharding.typed.javadsl.ShardedDaemonProcess;
import akka.japi.Pair;
import akka.persistence.query.Offset;
import akka.persistence.query.typed.EventEnvelope;
import akka.persistence.r2dbc.query.javadsl.R2dbcReadJournal;
import akka.projection.ProjectionBehavior;
import akka.projection.ProjectionId;
import akka.projection.eventsourced.javadsl.EventSourcedProvider;
import akka.projection.javadsl.ExactlyOnceProjection;
import akka.projection.javadsl.SourceProvider;
import akka.projection.r2dbc.javadsl.R2dbcProjection;

import java.util.List;
import java.util.Optional;
import akka.projection.r2dbc.R2dbcProjectionSettings;
import shopping.cart.repository.ItemPopularityRepository;

public final class ItemPopularityProjection {

    private ItemPopularityProjection() {
    }

    // tag::howto-read-side-without-role[]
    public static void init(
            ActorSystem<?> system,
            ItemPopularityRepository repository) {

        ShardedDaemonProcess.get(system)
                .initWithContext( // <1>
                        ProjectionBehavior.Command.class,
                        "ItemPopularityProjection",
                        4,
                        daemonContext -> {
                            List<Pair<Integer, Integer>> sliceRanges =
                                    EventSourcedProvider.sliceRanges(system, R2dbcReadJournal.Identifier(), daemonContext.totalProcesses());
                            Pair<Integer, Integer> sliceRange =
                                    sliceRanges.get(daemonContext.processNumber());
                            return ProjectionBehavior.create(createProjection(system, repository, sliceRange));
                        },
                        ShardedDaemonProcessSettings.create(system),
                        Optional.of(ProjectionBehavior.stopMessage()));
    }
    // end::howto-read-side-without-role[]

    private static ExactlyOnceProjection<Offset, EventEnvelope<ShoppingCart.Event>>
    createProjection(
            ActorSystem<?> system,
            ItemPopularityRepository repository,
            Pair<Integer, Integer> sliceRange) {

        int minSlice = sliceRange.first();
        int maxSlice = sliceRange.second();

        SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
                EventSourcedProvider.eventsBySlices(
                        system,
                        R2dbcReadJournal.Identifier(),
                        "ShoppingCart",
                        minSlice,
                        maxSlice);

        String slice = "carts-" + minSlice + "-" + maxSlice;
        Optional<R2dbcProjectionSettings> settings = Optional.empty();
        return R2dbcProjection.exactlyOnce( // <5>
                ProjectionId.of("ItemPopularityProjection", slice),
                settings,
                sourceProvider,
                () -> new ItemPopularityProjectionHandler(slice, repository),
                system);
    }
}
// end::projection[]
